Reduce work and allocations in dependency node

drewnoakes opened the pull request:
Pushing these small changes ahead of a larger PR based off this, to make the next PR easier to review.

---

Out of curiousity I loaded `Roslyn.sln`, performed a search in the solution, and counted the number of each kind of `DependencyModel` subclass created. Results:

| Type | Count |
|----|----:|
| `AssemblyDependencyModel` | 17,315 |
| `PackageDependencyModel` | 14,819 |
| `PackageDependencyModel` | 7,764 |
| `ProjectDependencyModel` | 3,756 |
| `AnalyzerDependencyModel` | 2,238 |
| `SdkDependencyModel` | 1,218 |
| `PackageAnalyzerDependencyModel` | 1,119 |
| `SharedProjectDependencyModel` | 10 |
| `SubTreeRootDependencyModel` | 5 |

Note this includes the change in 31813b78708fb9f498c58a2e5f8c31e6099a1b50 where root nodes are reused (hence the 5). That number would probably be just under 1,000 otherwise.

The gains here are modest, but they keep adding up.

jmarolf commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/MetadataExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            if (properties != null &&
+                properties.TryGetValue(key, out stringValue) &&
+                !string.IsNullOrEmpty(stringValue))
```
Why would we have a null or empty string in there to begin with?

drewnoakes replied:
I'm not sure why, but this check is pretty common amongst existing code that reads properties from dictionaries. Would `<EmptyElement></EmptyElement>` not produce an empty property?

jmarolf replied:
That still doesn't explain the null case for me

---
In reply to: [231937916](https://github.com/dotnet/project-system/pull/4242#discussion_r231937916) [](ancestors = 231937916)

jmarolf replied:
And if the property is an empty string why is it incorrect to return it?

---
In reply to: [231938983](https://github.com/dotnet/project-system/pull/4242#discussion_r231938983) [](ancestors = 231938983,231937916)

drewnoakes replied:
What would be better here?

jmarolf replied:
I don't know, I am moistly asking about the intended semantics of this method.  Its clear that if we have a value for a key that is null or an empty string we consider that the wrong result and return false, but it is unclear to me why 

---
In reply to: [231939349](https://github.com/dotnet/project-system/pull/4242#discussion_r231939349) [](ancestors = 231939349)

drewnoakes replied:
The intent of the `Get*` variants is that they return `null` if no sensible value could be obtained, and the caller can then `??` a default in. It's the more convenient form. You can `?? string.Empty` if you want.

I'll document these methods.

jmarolf replied:
Yes but if someone added an empty string or even null intentially to this dictionary why are those values not sensible? Essentially I am asking why this is wrong:

```C#
if (properties != null && properties.TryGetValue(key, out stringValue) {
   return true;
}
```

drewnoakes replied:
The decision was based on existing usages. Many test for empty strings and do something else if they're observed. This just makes that simpler. My understanding is that in MSBuild and empty string is equivalent to a null, as far as property values are concerned. Correct me if I'm mistaken.

jmarolf commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/MetadataExtensions.cs:
```diff
@@ -0,0 +1,80 @@
+
+            projectRuleSnapshot.Items.TryGetValue(itemSpec, out IImmutableDictionary<string, string> properties);
+
+            return properties;
```
Do we want to explicitly return null if `TryGetValue` fails?

drewnoakes replied:
If `TryGetValue` fails, `properties` is definitely `null`. Are you saying the code could be clearer?

jmarolf replied:
Yes, `properties` being bull is an implementation detail.  I assume we want to be explicit that we intend to return null

---
In reply to: [231938165](https://github.com/dotnet/project-system/pull/4242#discussion_r231938165) [](ancestors = 231938165)

drewnoakes replied:
I'll make it explicit.

drewnoakes commented:
@jmarolf I've addressed your feedback. Thanks.

@all is there a reason we can't bump the langversion to 7.3? We could use the `Enum` type constraint.

Pilchie commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependencyModel.cs:
```diff
@@ -9,45 +9,64 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.Models
     internal abstract class DependencyModel : IDependencyModel
     {
+        [Flags]
+        private enum DependencyFlags
```
Not sure if it would help due to field packing, but consider making this have an underlying type of `byte`?

jmarolf commented:
@Pilchie they have not. We either need newer CI machines or move our build to be completely independent of the VS compiler and targets

Pilchie commented:
> move our build to be completely independent of the VS compiler and targets

Looking forward to .NET Core 3 for WPF support for this part..
